src/Android/Avalonia.Android/Platform/SkiaPlatform/TopLevelImpl.cs

```diff
@@ -234,20 +200,40 @@ public override bool DispatchKeyEvent(KeyEvent? e)
                 return res ?? baseResult;
             }
 
-            void ISurfaceHolderCallback.SurfaceChanged(ISurfaceHolder holder, Format format, int width, int height)
+            public override void SurfaceChanged(ISurfaceHolder holder, Format format, int width, int height)
             {
-                var newSize = new PixelSize(width, height).ToSize(_tl.RenderScaling);
+                base.SurfaceChanged(holder, format, width, height);
+
+                var newSize = Size.ToSize(Scaling);
+                var newScaling = Scaling;
 
                 if (newSize != _oldSize)
                 {
                     _oldSize = newSize;
                     _tl.OnResized(newSize);
                 }
+                // ReSharper disable once CompareOfFloatsByEqualityOperator
+                if (newScaling != _oldScaling)
+                {
+                    _oldScaling =  newScaling;
+                    _tl.ScalingChanged?.Invoke(newScaling);
+                }
+            }
 
-                base.SurfaceChanged(holder, format, width, height);
+            public override void SurfaceRedrawNeeded(ISurfaceHolder holder)
+            {
+                // Compositor Renderer handles Paint event in-sync, which is perfect for sync SurfaceRedrawNeeded
+                _tl.Paint?.Invoke(new Rect(new Point(), Size.ToSize(Scaling)));
+                base.SurfaceRedrawNeeded(holder);
+            }
+
+            public override void SurfaceRedrawNeededAsync(ISurfaceHolder holder, IRunnable drawingFinished)
+            {
+                _tl.Compositor.RequestCompositionUpdate(drawingFinished.Run);
```

Comment:
I don't think that `RequestCompositionUpdate` is the best equivalent for "ImmediateRenderRequested, but async". But I think it's close enough for resizing and screen rotation.